fix(docs-infra): do not redirect disambiguated URLs (and other related infrastruture fixes)

gkalpak opened the pull request:
In #41788, logic was added to disambiguate case-insensitively equal docs paths/URLs. This process includes appending a `-\d+` suffix to some paths/URLs (for example, `/.../inject-1`). Unfortunately, some of the Firebase redirects configured in `firebase.json` would match these URLs and redirect them to non-existing paths.
Example failures: [stable][1], [next][2]

NOTE:
This was not picked up in the regular CI tests run for PRs, because the local devserver and the preview server used to test PRs do not support Firebase-like redirects.

This PR fixes this by ensuring these disambiguated paths/URLs are not matched by the redirect rules by checking whether the part of the suffix after the `-` contains any numeric digits. While this check is not ideal, it should be good enough for our purpose, since the legacy URLs that we do want to redirect contain suffixes such as `-class`, `-function` and thus no numeric digits.

This PR also includes related infrastructure/testing fixes to allow implementing the fix. See the individual commits for details.

[1]: https://circleci.com/gh/angular/angular/974345
[2]: https://circleci.com/gh/angular/angular/974346

mary-poppins commented:
You can preview 29eca5c at https://pr41842-29eca5c.ngbuilds.io/.

petebacondarwin reviewed (approved):
A few minor nits but nice refactoring.

This does beg the question though... at what point should we stop supporting a redirect? Some of these are 5+ years old, so it would seem very unlikely that they are still relevant. And in any case, they should probably link through to v2.angular.io or similar archive version for them to make sense anyway.

Also, I was wondering, regarding the last commit... rather than not redirecting URLs that end in `-\d+` how about we just enumerate all the endings that do need redirecting? E.g.`*-class`, `*-directive` etc. I think we must know what they are and there can't be that many of them.

gkalpak commented on aio/tools/firebase-test-utils/FirebaseRedirectSource.spec.ts:
```diff
@@ -2,188 +2,320 @@ import { FirebaseRedirectSource } from './FirebaseRedirectSource';
+          '/a/': undefined,
+          '/a/b/': undefined,
+          '/a/b/c/': undefined,
+        });
```
When using a regex, we don't really have rest groups.

gkalpak committed:
test(docs-infra): fix `create-example` tests on Windows

Previously, the tests in `create-example.spec.ts` made assertions using
some hard-coded absolute paths (something like `/foo/bar`). This caused
the tests to fail on Windows, where the absolute paths are prefixed with
the drive letter (something like `C:/foo/bar`).

This commit uses `path.resolve()` to ensure paths are converted to the
format used on the current OS.

gkalpak committed:
refactor(docs-infra): prepare `firebase-test-utils` for accepting regex-based redirects

Currently, the utilities for testing Firebase redirects assume that the
redirects are configured using the glob-based `source` property.
However, Firebase also supports configuring redirects using regular
expressions (via the `regex` property).
See the [Firebase docs][1] for more details.

This commit refactors the utilities in `firebase-test-utils/` to make it
easy to add support for such regex-based redirect configurations.

[1]: https://firebase.google.com/docs/hosting/full-config#redirects

gkalpak committed:
test(docs-infra): make it easier to identify the failing URL in `testFirebaseRedirection.spec.ts`

This commit includes the URL under test in the test description in
`testFirebaseRedirection.spec.ts` to make it easier to identify the
affected URL when a test fails.

It also avoids unnecessarily creating multiple `FirebaseRedirector`
instances by sharing instances between tests.

gkalpak committed:
test(docs-infra): update `firebase-test-utils` to support regex-based redirects

This commit updates the utilities in `firebase-test-utils/` to also
support testing Firebase redirects that are configured using regular
expressions (via the `regex` property).
See the [Firebase docs][1] for more details.

[1]: https://firebase.google.com/docs/hosting/full-config#redirects

gkalpak committed:
fix(docs-infra): do not redirect disambiguated URLs

In #41788, logic was added to disambiguate case-insensitively equal docs
paths/URLs. This process includes appending a `-\d+` suffix to some
paths/URLs (for example, `/.../inject-1`). Unfortunately, some of the
Firebase redirects configured in `firebase.json` would match these URLs
and redirect them to non-existing paths.
Example failures: [stable][1], [next][2]

NOTE:
This was not picked up in the regular CI tests run for PRs, because the
local devserver and the preview server used to test PRs do not support
Firebase-like redirects.

This commit fixes this by ensuring these disambiguated paths/URLs are
not matched by the redirect rules by checking whether the part of the
suffix after the `-` contains any numeric digits. While this check is
not ideal, it should be good enough for our purpose, since the legacy
URLs that we do want to redirect contain suffixes such as `-class`,
`-function` and thus no numeric digits.

[1]: https://circleci.com/gh/angular/angular/974345
[2]: https://circleci.com/gh/angular/angular/974346

gkalpak committed:
build(docs-infra): take disambiguated doc paths into account when generating the sitemap

In ##41788, the `disambiguateDocsPathsProcessor` was introduced to fix
an issue with case-insensitively equal paths. This processor may alter
the paths of some docs and thus their final URL in the app.

Previously, both the `disambiguateDocPathsProcessor` and the
`createSitemap` processor (which relies on the docs' computed paths to
generate the sitemap file) were configured to run before the
"rendering-docs" phase. However, this resulted in the
`disambiguateDocPathsProcessor`'s running after `createSitemap`, which
meant that the sitemap did not include the updated doc paths.

This commit fixes it by ensuring that the
`disambiguateDocPathsProcessor` is explicitly run before the
`createSitemap` processor, so that the latter will be able to take into
account any changes made by the former.

gkalpak committed:
fixup! test(docs-infra): update `firebase-test-utils` to support regex-based redirects

mary-poppins commented:
You can preview ce2e3b5 at https://pr41842-ce2e3b5.ngbuilds.io/.

gkalpak commented:
> at what point should we stop supporting a redirect? Some of these are 5+ years old, so it would seem very unlikely that they are still relevant.

True. Let's discuss this in the next meeting.
(We might need to figure out how to track when a redirect was last added/needed, but otherwise it is reasonable to time-box these redirects.)

> And in any case, they should probably link through to v2.angular.io or similar archive version for them to make sense anyway.

Not sure that is necessarily true. There are pros and cons in both approaches.

> rather than not redirecting URLs that end in -\d+ how about we just enumerate all the endings that do need redirecting? E.g.*-class, *-directive etc. I think we must know what they are and there can't be that many of them.

I am not sure what these suffixes are (nor how to find out reliably). But happy to go that way if there is a good way to figure them out. (It might also be unnecessary if we decide to drop support for these old redirects anyway.)

jessicajaniuk commented:
This is merged into `master` and `12.0.x` but not `11.2.x` due to conflicts. Please create a patch PR targeting `11.2.x`.

gkalpak commented:
For reference, here is the PR for patch: #41854
